File: tools/migrate-photos.py
#!/usr/bin/env python3
"""
Перенос фото из content/products_new/ в images/products/
Python-версия (без Node.js).

Запуск (из корня репозитория):
    python3 migrate-photos.py

При первом запуске установит библиотеку Pillow автоматически.
"""

import json
import os
import sys
import subprocess
from pathlib import Path

# Ставим Pillow если его нет
try:
    from PIL import Image, ImageOps
except ImportError:
    print("📦 Устанавливаю библиотеку Pillow для обработки картинок...")
    # Пробуем разные варианты установки (зависит от версии Python / ОС)
    install_attempts = [
        [sys.executable, "-m", "pip", "install", "--user", "Pillow"],
        [sys.executable, "-m", "pip", "install", "--user", "--break-system-packages", "Pillow"],
        [sys.executable, "-m", "pip", "install", "Pillow"],
    ]
    installed = False
    for cmd in install_attempts:
        try:
            subprocess.check_call(cmd, stderr=subprocess.DEVNULL)
            installed = True
            break
        except subprocess.CalledProcessError:
            continue
    if not installed:
        print("❌ Не удалось установить Pillow автоматически.")
        print("   Попробуй вручную в терминале:")
        print(f"   {sys.executable} -m pip install --user Pillow")
        sys.exit(1)
    from PIL import Image, ImageOps

ROOT = Path(__file__).resolve().parent
SRC_DIR = ROOT / "content" / "products_new"
PRODUCTS_DIR = ROOT / "content" / "products"
DST_DIR = ROOT / "images" / "products"

IMG_EXTS = {".jpg", ".jpeg", ".png", ".webp", ".heic"}


def list_images(folder: Path):
    """Возвращает список изображений в папке, отсортированных по имени."""
    if not folder.exists():
        return []
    files = [
        f for f in folder.iterdir()
        if f.is_file()
        and f.suffix.lower() in IMG_EXTS
        and not f.name.startswith(".")
    ]
    return sorted(files, key=lambda p: p.name)


def process_image(src: Path, dst: Path) -> bool:
    """Сжимает изображение в WebP 1000px максимум по длинной стороне."""
    try:
        with Image.open(src) as img:
            # поворот по EXIF
            img = ImageOps.exif_transpose(img)
            # конвертация в RGB (webp не любит палитровые режимы со сложной прозрачностью)
            if img.mode in ("RGBA", "LA"):
                bg = Image.new("RGB", img.size, (255, 255, 255))
                bg.paste(img, mask=img.split()[-1])
                img = bg
            elif img.mode != "RGB":
                img = img.convert("RGB")
            # ресайз
            img.thumbnail((1000, 1000), Image.LANCZOS)
            # сохранение
            img.save(dst, "WEBP", quality=85, method=6)
        return True
    except Exception as e:
        print(f"  ✗ Ошибка при обработке {src.name}: {e}")
        return False


def main():
    print("🔍 Читаю файлы товаров из content/products/...")

    if not PRODUCTS_DIR.exists():
        print(f"❌ Не найдена папка {PRODUCTS_DIR}")
        print("   Убедись что ты запустил скрипт из корня репозитория xpodium-russia")
        sys.exit(1)

    DST_DIR.mkdir(parents=True, exist_ok=True)

    product_files = sorted(PRODUCTS_DIR.glob("*.json"))
    total_processed = 0
    total_skipped = 0

    for product_path in product_files:
        with product_path.open(encoding="utf-8") as f:
            product = json.load(f)

        if "_new_folder" not in product:
            continue

        folders = [product["_new_folder"]] + product.get("_extra_folders", [])
        all_files = []
        for folder_name in folders:
            folder = SRC_DIR / folder_name
            all_files.extend(list_images(folder))

        slug = product.get("id", product_path.stem)
        name_ru = product.get("name_ru", "")

        if not all_files:
            print(f"⚠️  {slug}: в папках {folders} не найдено фото — пропускаю")
            total_skipped += 1
            # Всё равно убираем временные поля
            product.pop("_new_folder", None)
            product.pop("_extra_folders", None)
            product_path.write_text(
                json.dumps(product, ensure_ascii=False, indent=2), encoding="utf-8"
            )
            continue

        print(f"\n📦 {slug} ({name_ru}) — {len(all_files)} фото из {', '.join(folders)}")

        new_image_paths = []
        for i, src in enumerate(all_files):
            suffix = "" if i == 0 else f"-{i + 1}"
            dst_name = f"{slug}{suffix}.webp"
            dst = DST_DIR / dst_name
            if process_image(src, dst):
                new_image_paths.append(f"/images/products/{dst_name}")
                print(f"  ✓ {src.name} → {dst_name}")

        if new_image_paths:
            product["images"] = new_image_paths
            total_processed += 1

        # Убираем временные поля
        product.pop("_new_folder", None)
        product.pop("_extra_folders", None)

        product_path.write_text(
            json.dumps(product, ensure_ascii=False, indent=2), encoding="utf-8"
        )

    print("\n" + "━" * 40)
    print(f"✅ Обработано: {total_processed} товаров, пропущено: {total_skipped}")

    # Пересобираем products.json
    build_script = ROOT / "build-content.js"
    if build_script.exists():
        print("\n🔨 Пересобираю products.json...")
        try:
            subprocess.run(["node", str(build_script)], cwd=ROOT, check=True)
        except FileNotFoundError:
            # node нет — это ок, products.json всё равно обновится при деплое на Netlify
            print("   (node не установлен — products.json пересоберётся автоматически на Netlify)")
        except subprocess.CalledProcessError as e:
            print(f"   Ошибка: {e}")

    print("\n👉 Дальше:")
    print("   1. Открой GitHub Desktop — увидишь кучу изменений")
    print('   2. Commit message: "Обновление фото товаров"')
    print("   3. Commit to main → Push origin")
    print("   4. Через минуту сайт обновится")


if __name__ == "__main__":
    main()

File: tools/migrate-photos.js
#!/usr/bin/env node
/**
 * Перенос фото из content/products_new/ в images/products/
 *
 * Что делает:
 * 1. Читает все файлы из content/products/*.json (источник данных товаров)
 * 2. Для каждого товара с полем _new_folder — берёт фото из content/products_new/<folder>/
 * 3. Сжимает в WebP (1000px по длинной стороне, качество 85)
 * 4. Сохраняет как <slug>.webp, <slug>-2.webp, <slug>-3.webp и т.д.
 *    (порядок — по алфавиту имён файлов; 1.jpg, 2.jpg будут первыми)
 * 5. Обновляет поле "images" в JSON-файле товара, удаляет _new_folder и _extra_folders
 * 6. В конце запускает build-content.js чтобы обновить products.json
 *
 * Запуск (из корня репозитория):
 *    node migrate-photos.js
 *
 * Требует: npm install sharp (делает автоматически при первом запуске)
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Проверяем / устанавливаем sharp (библиотека для работы с картинками)
try {
  require.resolve('sharp');
} catch {
  console.log('Устанавливаю библиотеку sharp для обработки картинок (~10 сек)...');
  execSync('npm install sharp --no-save', { stdio: 'inherit' });
}
const sharp = require('sharp');

const ROOT = __dirname;
const SRC_DIR = path.join(ROOT, 'content', 'products_new');
const PRODUCTS_DIR = path.join(ROOT, 'content', 'products');
const DST_DIR = path.join(ROOT, 'images', 'products');

const IMG_EXTS = ['.jpg', '.jpeg', '.png', '.webp', '.heic'];

function listImages(folder) {
  if (!fs.existsSync(folder)) return [];
  return fs
    .readdirSync(folder)
    .filter(f => IMG_EXTS.includes(path.extname(f).toLowerCase()))
    .filter(f => !f.startsWith('.'))
    .sort();
}

async function processImage(srcPath, dstPath) {
  try {
    await sharp(srcPath)
      .rotate()
      .resize(1000, 1000, { fit: 'inside', withoutEnlargement: true })
      .webp({ quality: 85 })
      .toFile(dstPath);
    return true;
  } catch (e) {
    console.error(`  ✗ Ошибка при обработке ${path.basename(srcPath)}: ${e.message}`);
    return false;
  }
}

(async () => {
  console.log('🔍 Читаю файлы товаров из content/products/...');

  if (!fs.existsSync(DST_DIR)) fs.mkdirSync(DST_DIR, { recursive: true });

  const productFiles = fs.readdirSync(PRODUCTS_DIR).filter(f => f.endsWith('.json'));

  let totalProcessed = 0;
  let totalSkipped = 0;

  for (const productFile of productFiles) {
    const productPath = path.join(PRODUCTS_DIR, productFile);
    const product = JSON.parse(fs.readFileSync(productPath, 'utf-8'));

    if (!product._new_folder) continue;

    const folders = [product._new_folder, ...(product._extra_folders || [])];
    const allFiles = [];
    for (const folder of folders) {
      const src = path.join(SRC_DIR, folder);
      const files = listImages(src);
      for (const f of files) {
        allFiles.push(path.join(src, f));
      }
    }

    const slug = product.id || productFile.replace('.json', '');

    if (!allFiles.length) {
      console.log(`⚠️  ${slug}: в папках ${folders.join(', ')} не найдено фото — пропускаю`);
      totalSkipped++;
      continue;
    }

    console.log(`\n📦 ${slug} (${product.name_ru}) — ${allFiles.length} фото из ${folders.join(', ')}`);

    const newImageNames = [];
    for (let i = 0; i < allFiles.length; i++) {
      const srcPath = allFiles[i];
      const suffix = i === 0 ? '' : `-${i + 1}`;
      const dstName = `${slug}${suffix}.webp`;
      const dstPath = path.join(DST_DIR, dstName);
      const ok = await processImage(srcPath, dstPath);
      if (ok) {
        newImageNames.push(`/images/products/${dstName}`);
        console.log(`  ✓ ${path.basename(srcPath)} → ${dstName}`);
      }
    }

    if (newImageNames.length) {
      product.images = newImageNames;
      totalProcessed++;
    }

    delete product._new_folder;
    delete product._extra_folders;

    // Сохраняем обновлённый файл товара
    fs.writeFileSync(productPath, JSON.stringify(product, null, 2), 'utf-8');
  }

  console.log('\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
  console.log(`✅ Обработано: ${totalProcessed} товаров, пропущено: ${totalSkipped}`);

  // Пересобираем products.json
  console.log('\n🔨 Пересобираю products.json...');
  try {
    execSync('node build-content.js', { stdio: 'inherit' });
  } catch (e) {
    console.error('Ошибка при сборке products.json:', e.message);
  }

  console.log('\n👉 Дальше:');
  console.log('   1. Открой GitHub Desktop');
  console.log('   2. Увидишь кучу изменений (новые фото + обновлённые JSON)');
  console.log('   3. Commit message: "Обновление фото товаров"');
  console.log('   4. Commit to main → Push origin');
  console.log('   5. Через минуту сайт обновится');
})();
